### ISSUE TITLE
Unable to connect to elasticsearch instance running on non-root path

### ISSUE THREAD

**Issue by karatekaneen** [elastic/elasticsearch-js#1709]
## 🐛 Bug Report

I'm unable to connect to elasticsearch instance running on non-root path, in my case `/elastic`. This is due to [this function](https://github.com/nodejs/undici/blob/966a54ff80993573d1baab5999ce258ccc0e7ec9/lib/core/util.js#L124-L132) in Undici that does not allow non-root paths. Might be related to #1625 but not sure so opening a new issue.

## To Reproduce

```js
const client = new Client({
  node: "http://secret.com/elastic",
  auth: {
    username: "admin",
    password: "admin",
  },
});
```

## Expected behavior

I would love to be able to connect to an Elasticsearch instance that is hidden behind a k8s ingress.

Output:

```js
InvalidArgumentError: invalid url
    at Object.parseOrigin (/workspace/node_modules/undici/lib/core/util.js:83:11)
    at new Pool (/workspace/node_modules/undici/lib/pool.js:62:23)
    at new Connection (/workspace/node_modules/@elastic/transport/src/connection/UndiciConnection.ts:113:17)
    at WeightedConnectionPool.createConnection (/workspace/node_modules/@elastic/transport/src/pool/BaseConnectionPool.ts:133:24)
    at WeightedConnectionPool.addConnection (/workspace/node_modules/@elastic/transport/src/pool/BaseConnectionPool.ts:158:53)
    at new Client (/workspace/node_modules/@elastic/elasticsearch/src/client.ts:231:27)
    at /workspace/src/server.ts:26:18
    at Generator.next (<anonymous>)
    at /workspace/src/server.ts:8:71
    at new Promise (<anonymous>)
    at __awaiter (/workspace/src/server.ts:4:12)
    at init (/workspace/src/server.ts:23:12)
    at Object.<anonymous> (/workspace/src/index.ts:3:5)
    at Module._compile (node:internal/modules/cjs/loader:1105:14)
    at Module.m._compile (/workspace/node_modules/ts-node/src/index.ts:1455:23)
    at Module._extensions..js (node:internal/modules/cjs/loader:1159:10) {
  code: 'UND_ERR_INVALID_ARG'
}
```

## Your Environment

- *node version*: 18
- `@elastic/elasticsearch` 8.2.1
- *os*: Linux

**Comment by sheikalthaf** [comment]
@karatekaneen  I'm too facing the same issue. Any workaround here?

**Comment by karatekaneen** [comment]
@sheikalthaf No. 
Fortunately for me I experienced this in a small proof-of-concept so I just did the raw http-request via axios instead of using the client lib in the meantime

**Comment by ahmed-hm** [comment]
I'm facing the same issue too in latest v8.4

**Comment by vdusauto** [comment]
We were facing the same issue and have been able to go around it.
Here is how we did it:

1. When creating client options, do:

`const clientOptions = {
      node: elasticSearchBaseUrl,
      Transport: TransportIncludingSubPath,
    } as ClientOptions;`
    
2. Define the class TransportIncludingSubPath as follows:

`export class TransportIncludingSubPath extends ElasticTransport {
  async request(
    params: TransportRequestParams,
    options?:
      | TransportRequestOptionsWithOutMeta
      | TransportRequestOptionsWithMeta,
  ): Promise<any> {
    params.path = `${elasticSearchUrlPrefix}${params.path}`;
    return super.request(params, options);
  }
}`

This should help you use the 8.x version of Elastic without the undici error

Hope it helps !

**Comment by alkaaf** [comment]
this might be help, i must dig deeper to create custom-like request to the ES server. Elasticsearch provide Transport Class to modify anyting before the request done to the server. this is common when the ES sit behind gateway or ingress controller (i am using nginx). i don't know why ES didn't create easier method to do this. Here is the complete code:

```javascript
const {Client} = require('@elastic/elasticsearch');
const {Transport} = require('@elastic/transport')
let baseUrl = "https://<your es base url>/"
let path = "<your es path ex: /es/my-es>"

// then create a class, that extends Transport class to modify the path
class MTransport extends Transport {
    request(params, options, callback) {
        params.path = path + params.path // <- append the path right here
        return super.request(params, options, callback)
    }
}

// and finally put the extended class on the options.
const client = new Client({
    node: baseUrl,
    Transport: MTransport,
    auth: {
        username: "xxxx",
        password: "secret"
    }
});
```

**Comment by JoshMock** [comment]
@vdusauto's recommendation seems like the best way to solve this problem, short of adding path prefix support directly to the client or `@elastic/transport`. I'm going to close this for now, but please let us know if this continues to be a challenge.

**Comment by will-hu-0** [comment]
> We were facing the same issue and have been able to go around it. Here is how we did it:
> 
> 1. When creating client options, do:
> 
> `const clientOptions = { node: elasticSearchBaseUrl, Transport: TransportIncludingSubPath, } as ClientOptions;`
> 
> 2. Define the class TransportIncludingSubPath as follows:
> 
> `export class TransportIncludingSubPath extends ElasticTransport { async request( params: TransportRequestParams, options?: | TransportRequestOptionsWithOutMeta | TransportRequestOptionsWithMeta, ): Promise<any> { params.path = elasticSearchUrlPrefix + params.path; return super.request(params, options); } }`
> 
> This should help you use the 8.x version of Elastic without the undici error
> 
> Hope it helps !

Thanks for this, you saved my day, @vdusauto .
I did some code a bit different from yours to make it working in my Typescript project:

1. In es_client.ts, I use `Transport`, not `ElasticTransport` from what you mentioned...

```
import {
  Client,
  Transport,
  TransportRequestOptionsWithMeta,
  TransportRequestOptionsWithOutMeta,
  TransportRequestParams,
  type ClientOptions
} from "@elastic/elasticsearch";

export class TransportIncludingSubPath extends Transport {
  async request(
    params: TransportRequestParams,
    options?: TransportRequestOptionsWithOutMeta | TransportRequestOptionsWithMeta
  ): Promise<any> {
    params.path = elasticSearchUrlPrefix + params.path;
    return super.request(params, options);
  }
}
... ...
```

2. in `tsconfig.json`, the `compilerOptions.target` needs to set to "ES2017" other than "ES5"...

Using "ES5" always returns errors like 
```
TypeError: Class constructor Transport cannot be invoked without 'new'
    at new TransportIncludingSubPath (/Users/will/Work/zComm/rag-utilities/build/clients/elasticsearch-client.js:96:42)
    at new Client (/Users/will/Work/zComm/rag-utilities/node_modules/@elastic/elasticsearch/lib/client.js:230:26)
    at elasticsearchClient (/Users/will/Work/zComm/rag-utilities/build/clients/elasticsearch-client.js:132:12)
```
Some explanation shows up from [here](https://stackoverflow.com/questions/51860043/javascript-es6-typeerror-class-constructor-client-cannot-be-invoked-without-ne).

**Comment by margaretjgu** [comment]
This has been fixed in 8.19.2 
https://github.com/elastic/elasticsearch-js/issues/3342
